src/explorer/explorer.ts

```diff
@@ -2,23 +2,101 @@ import * as vscode from "vscode";
+import { config, projectsExplorerProvider } from "../extension";
 import { WorkspaceNode } from "./models/workspaceNode";
+import { outputChannel } from "../utils";
+import { DocumentContentProvider } from "../providers/DocumentContentProvider";
+
+/** Get the URI for this leaf node */
+export function getLeafNodeUri(node: NodeBase, forceServerCopy = false): vscode.Uri {
+  if (node.workspaceFolder == undefined) {
+    // Should only be the case for leaf nodes in the projects explorer
+    // that are children of an extra server namepsace node
+    return DocumentContentProvider.getUri(
+      node.fullName,
+      undefined,
+      undefined,
+      true,
+      node.workspaceFolderUri,
+      forceServerCopy
+    );
+  } else {
+    return DocumentContentProvider.getUri(
+      node.fullName,
+      node.workspaceFolder,
+      node.namespace,
+      undefined,
+      undefined,
+      forceServerCopy
+    );
+  }
+}
 
-// The vscode-objectscript.explorer.open command is not listed in settings.json as a contribution because it only gets invoked
-//  from the user's click on an item in the ObjectScriptExplorerProvider tree.
-// It serves as a proxy for the vscode.open command, detecting two opens of the same item in quick succession
-//  and treating the second of these as a non-preview open.
+/**
+ * The vscode-objectscript.explorer.open command is not listed in settings.json as a contribution because it only gets invoked
+ * from the user's click on an item in the `ObjectScriptExplorerProvider` or `ProjectsExplorerProvider` tree.
+ * It serves as a proxy for `vscode.window.showTextDocument()`, detecting two opens of the same item in quick succession
+ * and treating the second of these as a non-preview open.
+ */
 export function registerExplorerOpen(explorerProvider: ObjectScriptExplorerProvider): vscode.Disposable {
-  return vscode.commands.registerCommand("vscode-objectscript.explorer.open", async function (uri: vscode.Uri) {
-    let usePreview = <boolean>vscode.workspace.getConfiguration("workbench.editor").get("enablePreview");
-
-    if (usePreview) {
-      usePreview = !wasDoubleClick(uri, explorerProvider);
+  return vscode.commands.registerCommand(
+    "vscode-objectscript.explorer.open",
+    async function (uri: vscode.Uri, project?: string, fullName?: string) {
+      let usePreview = <boolean>vscode.workspace.getConfiguration("workbench.editor").get("enablePreview");
+
+      if (usePreview) {
+        usePreview = !wasDoubleClick(uri, explorerProvider);
+      }
+
+      try {
+        await vscode.window.showTextDocument(uri, { preview: usePreview });
+      } catch (error) {
+        if (project && fullName) {
+          // This project item no longer exists on the server
+          // Ask the user if they would like to remove it from the project
+          const remove = await vscode.window.showErrorMessage(
+            `Document '${fullName}' does not exist on the server. Remove it from project '${project}'?`,
```

Comment:
I've been seeing this occur unexpectedly. Here's a screenshot example:

![image](https://user-images.githubusercontent.com/6726799/163175071-054cab0a-fbdf-485a-886e-09c7e12cbfdc.png)

You can see that I was able to open the class from within the project, but almost immediately after it loaded I got the alert saying it didn't exist on the server.